File: sdks/risk-client-typescript/src/webhooks.ts
import * as crypto from 'crypto';
import { ClientOptions, Subscription } from './types';
import { fetchWithRetry } from './http';
import { resolveEnv } from './env';

export class WebhooksChannel {
  private readonly baseUrl: string;

  constructor(private readonly options: ClientOptions) {
    this.baseUrl = resolveEnv(options.environment).restBaseUrl;
  }

  async subscribe(callbackUrl: string, eventFilter: string): Promise<Subscription> {
    return fetchWithRetry<Subscription>(
      `${this.baseUrl}/webhook/register`,
      {
        method: 'POST',
        body: JSON.stringify({
          callbackUrl,
          events: eventFilter.split(','),
        }),
      },
      this.options,
    );
  }

  async unsubscribe(subscriptionId: string): Promise<void> {
    await fetchWithRetry<void>(
      `${this.baseUrl}/webhook/unregister/${subscriptionId}`,
      { method: 'POST', body: JSON.stringify({}) },
      this.options,
    );
  }

  async list(): Promise<Subscription[]> {
    return fetchWithRetry<Subscription[]>(
      `${this.baseUrl}/webhook/subscriptions`,
      { method: 'GET' },
      this.options,
    );
  }

  /**
   * Verify an inbound webhook payload against its HMAC-SHA256 signature.
   * Uses a constant-time comparison to prevent timing attacks.
   */
  verify(payload: Buffer | string, signature: string, signingSecret: string): boolean {
    const hmac = crypto
      .createHmac('sha256', signingSecret)
      .update(payload)
      .digest('hex');

    try {
      return crypto.timingSafeEqual(
        Buffer.from(hmac,      'hex'),
        Buffer.from(signature, 'hex'),
      );
    } catch {
      return false;
    }
  }
}

File: sdks/risk-client-go/webhooks.go
package riskclient

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

// WebhooksClient manages webhook subscriptions.
type WebhooksClient struct {
	http    *jsonHTTP
	baseURL string
}

func newWebhooksClient(cfg Config, h *jsonHTTP) *WebhooksClient {
	return &WebhooksClient{http: h, baseURL: envMap[cfg.Environment].restBaseURL}
}

// Subscribe registers callbackURL for the given event filter (e.g. "DECLINE,REVIEW").
func (w *WebhooksClient) Subscribe(ctx context.Context, callbackURL, eventFilter string) (*Subscription, error) {
	body := map[string]any{
		"callbackUrl": callbackURL,
		"events":      splitFilter(eventFilter),
	}
	var sub Subscription
	if err := w.http.postJSON(ctx, w.baseURL+"/webhook/register", body, &sub); err != nil {
		return nil, err
	}
	return &sub, nil
}

// Unsubscribe removes a subscription by ID.
func (w *WebhooksClient) Unsubscribe(ctx context.Context, subscriptionID string) error {
	url := fmt.Sprintf("%s/webhook/unregister/%s", w.baseURL, subscriptionID)
	return w.http.postJSON(ctx, url, map[string]any{}, nil)
}

// List returns all subscriptions for this API key.
func (w *WebhooksClient) List(ctx context.Context) ([]Subscription, error) {
	var subs []Subscription
	if err := w.http.getJSON(ctx, w.baseURL+"/webhook/subscriptions", &subs); err != nil {
		return nil, err
	}
	return subs, nil
}

// Verify checks that payload matches the provided HMAC-SHA256 signature.
// Uses constant-time comparison.
func (w *WebhooksClient) Verify(payload []byte, signature, signingSecret string) bool {
	mac := hmac.New(sha256.New, []byte(signingSecret))
	mac.Write(payload)
	expected := hex.EncodeToString(mac.Sum(nil))

	sigBytes, err := hex.DecodeString(signature)
	if err != nil {
		return false
	}
	expBytes, _ := hex.DecodeString(expected)
	return hmac.Equal(expBytes, sigBytes)
}

func splitFilter(filter string) []string {
	if filter == "" {
		return nil
	}
	var parts []string
	start := 0
	for i := 0; i < len(filter); i++ {
		if filter[i] == ',' {
			parts = append(parts, filter[start:i])
			start = i + 1
		}
	}
	parts = append(parts, filter[start:])
	return parts
}
